### ISSUE TITLE
Add preparation mechanism

### ISSUE THREAD

**Issue by adrhill** [adrhill/SparseConnectivityTracer.jl#218]
SCT could benefit from the addition of a preparation mechanism.
The following points primarily apply to local patterns, but don't harm global pattern detection.

## 1. Heuristics for the selection of pattern representations
The [benchmarks](https://github.com/adrhill/SparseConnectivityTracer.jl/pull/211#issuecomment-2457887211) in #211 have (once again) shown that there is no free lunch when it comes to the best internal pattern representation, as it depends on (a) the sparsity of the given function and (b) the size of index sets over the entire compute graph. The most-user friendly solution to this problem would be to introduce a heuristic, e.g. on input size[^1]. 

Since such a run-time selection of internal types is by definition type unstable, this heuristic should be moved outside of any hot loops, e.g. by moving it to a type-unstable preparation mechanism.

## 2. Avoid repeated input enumeration
Local pattern detection methods currently enumerate index sets on every call. This could be avoided by returning a reusable buffer when calling the preparation mechanism.

This feature will get more important in the context of reusable chunked index sets (see #194).

## 3. Chunked Hessian detection?
A preparation mechanism for Hessian patterns that first computes a global hessian pattern might be a possible avenue for chunked local Hessian detection.

## 4. Composability with DI
Ideally, SCT's preparation mechanism should be designed with DI's preparation of `AutoSparse` backends in mind, as the former should be called within the latter (CC @gdalle).

[^1]: In theory, the heuristic for local pattern detection could also call global pattern detection methods to get an upper estimate of the sparsity of a Jacobian/Hessian in addition to the input size.

**Comment by adrhill** [comment]
Since the preparation is only useful for local sparsity detection, we might be able to handle it through the `TracerLocalSparsityDetector` struct by adding constructors that take global sparsity patterns.

**Comment by gdalle** [comment]
I don't understand what you just said ^^

**Comment by adrhill** [comment]
We could allow for `detector = TracerLocalSparsityDetector(global_pattern)` to be smarter than the regular `TracerLocalSparsityDetector()`.

**Comment by adrhill** [comment]
We already have #135?

**Comment by gdalle** [comment]
Sure but gradients are still allocated

**Comment by gdalle** [comment]
Right. Then we can't provide a global pattern when the detector is created because we don't know on which inputs it will be applied. Maybe we could do a self-preparing local detector which chooses the representation on the first call?

**Comment by adrhill** [comment]
> Then we can't provide a global pattern when the detector is created because we don't know on which inputs it will be applied. 

This optional preparation mechanism would have assume that the provided pattern is computed for the same input dimensions.

> Maybe we could do a self-preparing local detector which chooses the representation on the first call?

This was my original idea, see "heuristic". The most informative preparation heuristic would be a one-time computation of the global sparsity pattern. However, that might not always work on branching code. Therefore my suggestion to support a generic `TracerLocalSparsityDetector(global_pattern)`. Something like `TracerLocalSparsityDetector(input_length=128, sparsity=0.42)` might also work well, even though it doesn't help with point 3.

**Comment by gdalle** [comment]
The heuristic could also just be a local sparsity pattern. What's wrong with that?

**Comment by adrhill** [comment]
That isn't more useful than `input_length=128, sparsity=0.42` since it is not conservative. Point 3 on chunked Hessians would not be possible, since more terms might interact than the non-conservative pattern reveals.

**Comment by gdalle** [comment]
For a performance heuristic, you don't really care if it's not conservative. You only need a ballpark estimate of how many elements your set will typically hold
